### PR TITLE
routing: return structured error for send to route

### PR DESCRIPTION
**Pull request by joostjager**
This PR fixes a bug where in case of non-final errors, no structured error was returned for the routerrpc SendToRoute.

Furthermore it fixes error reporting from findPath and adds missing channel updates fields to the response.

### CODE REVIEW

**Comment by joostjager** [comment]
Not sure if this fix is good enough. It fixes the new `SendToRoute` rpc, but also changes the response of the main rpc server one. Is that a problem? (Because it was already string based anyway). Scripts that try to parse that output may break.

**Comment by Roasbeef** [comment]
>  but also changes the response of the main rpc server one

Is a portion of this diff missing? I don't see where it changes the response for the existing main rpc server.

**Comment by alexbosworth** [comment]
lgtm

**Comment by joostjager** [comment]
>Is a portion of this diff missing? I don't see where it changes the response for the existing main rpc server.

It also calls into `SendToRoute`.

But the main rpc errors were already pretty inconsistent, so not sure if it really matters.

**Comment by joostjager** [comment]
Looked into the main rpc `SendToRoute` output a bit more. This are three errors generated with v0.6.1-beta:

1. `FinalExpiryTooSoon`

2. `unable to route payment to destination: UnknownNextPeer`

3. `unable to route payment to destination: FeeInsufficient(htlc_amt==5000000 mSAT, update=(lnwire.ChannelUpdate) {\n Signature: (lnwire.Sig) (len=64 cap=64) {\n  00000000  94 cb 01 02 ce 76 41 92  e1 43 b6 f2 80 c6 72 ec  |.....vA..C....r.|\n  00000010  56 bd f9 15 4d ec 54 d1  fc c8 9a 51 09 eb dd 9a  |V...M.T....Q....|\n  00000020  6d 57 84 39 e2 f8 7d 71  b1 c2 63 1e 27 e3 da f6  |mW.9..}q..c.'...|\n  00000030  e0 7d 77 2b 1f 23 0a 21  94 6c 7a fa ea 64 1d c1  |.}w+.#.!.lz..d..|\n },\n ChainHash: (chainhash.Hash) (len=32 cap=32) 0f9188f13cb7b2c71f2a335e3a4fc328bf5beb436012afca590b1a11466e2206,\n ShortChannelID: (lnwire.ShortChannelID) 2386:1:0,\n Timestamp: (uint32) 1558680311,\n MessageFlags: (lnwire.ChanUpdateMsgFlags) 00000001,\n ChannelFlags: (lnwire.ChanUpdateChanFlags) 00000001,\n TimeLockDelta: (uint16) 144,\n HtlcMinimumMsat: (lnwire.MilliSatoshi) 1000 mSAT,\n BaseFee: (uint32) 1000000,\n FeeRate: (uint32) 1000000,\n HtlcMaximumMsat: (lnwire.MilliSatoshi) 9900000000 mSAT,\n ExtraOpaqueData: ([]uint8) \u003cnil\u003e\n}\n`

The final error 1 isn't prefix with "unable to route payment to destination", while the non-final ones are.

After this pr, the main rpc errors become:
1. `FinalExpiryTooSoon`

2. `UnknownNextPeer`

3. `FeeInsufficient(htlc_amt==5000000 mSAT, update=(lnwire.ChannelUpdate) {\n Signature: (lnwire.Sig) (len=64 cap=64) {\n  00000000  79 2d c5 ce c0 d0 c8 b5  ec 41 7d 5c fa 23 5b 25  |y-.......A}\\.#[%|\n  00000010  5a 84 e8 cc 2a 01 31 bf  61 2b 98 92 2a bd 00 5b  |Z...*.1.a+..*..[|\n  00000020  7b 0d 80 14 53 a0 02 64  fd 6e 93 d4 55 00 93 85  |{...S..d.n..U...|\n  00000030  95 c9 32 28 83 61 e3 c3  34 3d ff aa 3f cc 3f 08  |..2(.a..4=..?.?.|\n },\n ChainHash: (chainhash.Hash) (len=32 cap=32) 0f9188f13cb7b2c71f2a335e3a4fc328bf5beb436012afca590b1a11466e2206,\n ShortChannelID: (lnwire.ShortChannelID) 2368:1:0,\n Timestamp: (uint32) 1558680311,\n MessageFlags: (lnwire.ChanUpdateMsgFlags) 00000001,\n ChannelFlags: (lnwire.ChanUpdateChanFlags) 00000001,\n TimeLockDelta: (uint16) 144,\n HtlcMinimumMsat: (lnwire.MilliSatoshi) 1000 mSAT,\n BaseFee: (uint32) 1000000,\n FeeRate: (uint32) 1000000,\n HtlcMaximumMsat: (lnwire.MilliSatoshi) 9900000000 mSAT,\n ExtraOpaqueData: ([]uint8) \u003cnil\u003e\n}\n`

**Comment by alexbosworth** [comment]
changes lgtm

**Comment by halseth** [comment]
Needs rebase!

File: routing/router.go [hunk 1]
```diff
@@ -1380,6 +1375,9 @@ func (r *ChannelRouter) FindRoute(source, target route.Vertex,
 		},
 		restrictions, source, target, amt,
 	)
+	if err != nil {
```
Comment (halseth) [inline routing/router.go]: arguably this could be put in its own PR that we can merge immediately :P

File: lnrpc/routerrpc/router.proto [hunk 2]
```diff
@@ -164,50 +164,85 @@ message Failure {
...
-	// satoshi.
-	uint32 fee_rate = 9;
+    /**
+    signature is used to validate the announced data and prove the ownership of
+    node id.
+    */
+    bytes signature = 1;
+
+    /**
+    chain_hash denotes the target chain that this channel was opened within.
```
[the remark this answers is not in the document]
Comment (joostjager) [reply lnrpc/routerrpc/router.proto]: Ah, that is the reason we avoid this. I was wondering for some time now why.

File: routing/router.go (hunk 1 above)
Comment (joostjager) [reply routing/router.go]: there is no such thing as "merge immediately" in `lnd` :)

File: lnrpc/routerrpc/router.proto (hunk 2 above)
[the remark this answers is not in the document]
Comment (joostjager) [reply lnrpc/routerrpc/router.proto]: Fixed

**Comment by joostjager** [comment]
@wpaulino @halseth ptal

File: routing/router.go (hunk 1 above)
Comment (halseth) [reply routing/router.go]: 😆

File: routing/payment_lifecycle.go [hunk 3]
```diff
@@ -12,6 +12,20 @@ import (
 	"github.com/lightningnetwork/lnd/routing/route"
 )
 
+// errNoRoute is returned when all routes from the payment session have been
+// attempted.
+type errNoRoute struct {
+	// lastError is the last attempt error, if at least one attempt has been
```
Comment (halseth) [inline routing/payment_lifecycle.go]: s/last attempt error/error encountered during the last attempt

File: routing/payment_lifecycle.go [hunk 4]
```diff
@@ -12,6 +12,20 @@ import (
 	"github.com/lightningnetwork/lnd/routing/route"
 )
 
+// errNoRoute is returned when all routes from the payment session have been
+// attempted.
+type errNoRoute struct {
```
Comment (halseth) [inline routing/payment_lifecycle.go]: Good solution 👍

File: routing/payment_lifecycle.go [hunk 5]
```diff
@@ -12,6 +12,20 @@ import (
...
+	// lastError is the last attempt error, if at least one attempt has been
+	// made.
+	lastError error
```
Comment (halseth) [inline routing/payment_lifecycle.go]: Hm, can we make this of type `ForwardingError` now?

File: routing/payment_lifecycle.go (hunk 3 above)
Comment (joostjager) [reply routing/payment_lifecycle.go]: Fixed

File: routing/payment_lifecycle.go (hunk 5 above)
Comment (joostjager) [reply routing/payment_lifecycle.go]: Yes that is possible. It was already possible before. Fixed.

**Comment by joostjager** [comment]
@halseth ptal

File: routing/payment_lifecycle.go [hunk 6]
```diff
@@ -37,7 +37,7 @@ type paymentLifecycle struct {
 	finalCLTVDelta uint16
 	attempt        *channeldb.PaymentAttemptInfo
 	circuit        *sphinx.Circuit
-	lastError      error
+	lastError      *htlcswitch.ForwardingError
```
Comment (halseth) [inline routing/payment_lifecycle.go]: should be done for `errNoRoute` error also. Can also squash the two commits together.

Comment (joostjager) [reply routing/payment_lifecycle.go]: yes, good point. fixed and squashed.

**Comment by joostjager** [comment]
ptal @halseth

File: routing/router_test.go [hunk 7]
```diff
@@ -3136,3 +3136,107 @@ func TestRouterPaymentStateMachine(t *testing.T) {
...
+	if err != nil {
+		t.Fatalf("unable to create graph: %v", err)
+	}
+
+	const startingBlockHeight = 101
+
+	ctx, cleanUp, err := createTestCtxFromGraphInstance(startingBlockHeight,
+		testGraph)
+
+	defer cleanUp()
```
[the remark this answers is not in the document]
Comment (joostjager) [reply routing/router_test.go]: It seems incorrect, fixed

File: routing/payment_lifecycle.go [hunk 8]
```diff
@@ -349,6 +373,5 @@ func (p *paymentLifecycle) handleSendError(sendErr error) error {
 	}
 
 	// We get ready to make another payment attempt.
-	p.lastError = sendErr
```
Comment (halseth) [inline routing/payment_lifecycle.go]: nit: comment above this line can be removed also.

File: routing/router.go [hunk 9]
```diff
@@ -1658,7 +1658,18 @@ func (r *ChannelRouter) SendToRoute(hash lntypes.Hash, route *route.Route) (
 	// Since this is the first time this payment is being made, we pass nil
 	// for the existing attempt.
 	preimage, _, err := r.sendPayment(nil, payment, paySession)
-	return preimage, err
+	if err != nil {
+		// SendToRoute should return a structured error. In case the provided
+		// route fails, payment lifecycle will return a noRouteError with the
+		// structured error embedded.
+		if noRouteError, ok := err.(errNoRoute); ok {
+			return lntypes.Preimage{}, noRouteError.lastError
```
Comment (halseth) [inline routing/router.go]: hm, what happens if `lastError` is nil here?

File: routing/router_test.go [hunk 10]
```diff
@@ -3136,3 +3136,107 @@ func TestRouterPaymentStateMachine(t *testing.T) {
...
+		symmetricTestChannel("b", "c", chanCapSat, &testChannelPolicy{
+			Expiry:  144,
+			FeeRate: 400,
+			MinHTLC: 1,
+			MaxHTLC: lnwire.NewMSatFromSatoshis(chanCapSat),
+		}, 2),
+	}
+
+	testGraph, err := createTestGraphFromChannels(testChannels)
+	defer testGraph.cleanUp()
```
Comment (halseth) [inline routing/router_test.go]: should be moved to after error check

File: routing/router.go (hunk 9 above)
Comment (halseth) [reply routing/router.go]: Should be possible to test by mocking a payment session that finds no route on first request.

Comment (joostjager) [reply routing/router.go]: It should never happen, because SendToRoute always provides a route in the payment session. Added nil check to prevent crash.

**Comment by joostjager** [comment]
@halseth ptal
